sdk/core/core-amqp/src/auth/tokenProvider.ts

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AccessToken, NamedKeyCredential, SASCredential } from "@azure/core-auth";
+import jssha from "jssha";
+import { isObjectWithProperties } from "../util/typeGuards";
+
+/**
+ * A TokenProvider provides an alternative to TokenCredential for providing an `AccessToken`.
+ * @hidden
+ */
+export interface TokenProvider {
```

Comment:
I went with the name `TokenProvider` because:
1. `TokenCredential` was already taken.
2. It _provides_ a token.

I am open to changing the name if someone has a better one, though note that customers shouldn't see this interface, it's only used internally by event hubs and service bus.

Reply:
We have different treatments for when a token is SAS vs JWT and this TokenProvider seems to be only for the SAS case. So, I would recommend naming it so to avoid confusion that this may one day be applicable to tokens from AAD

SasTokenProvider ?

Reply:
Or we can keep the name `TokenProvider` here for the interface and have a single class `SasTokenProvider` instead of having two for NamedKey and Sas. Then in EH & SB, one can do an `instanceOf SasTokenProvider` check to decide the different treatments for SAS token vs AAD token

Reply:
> Then in EH & SB, one can do an instanceOf SasTokenProvider

I really like to stay away from instanceof checks in JS if possible, since it can fail in some weird edge cases (such as if multiple copies of the library are imported indirectly).

> have a single class SasTokenProvider instead of having two for NamedKey and Sas.

I think it can make sense to rename it to `SasTokenProvider` as long as we don't plan to add other 'TokenProvider' scenarios :smile: I'm not sure I see the value in combining the two TokenProvider implementations into a single class since both their internal state and how they 'calculate' a token are completely different.

Reply:
On further reflection, I actually like having a single `SasTokenProvider` class, since both current implementations provide a Sas token. Updating...

Reply:
woohoo!

Reply:
When you're right, you're right!

Reply:
Updated! And rebased against master so no more merge conflicts.